designs/2024-support-ts-config-files/README.md

```diff
@@ -0,0 +1,335 @@
+- Repo: eslint/eslint
+- Start Date: 2024-03-09
+- RFC PR: <https://github.com/eslint/rfcs/pull/117>
+- Authors: [Arya Emami](https://github.com/aryaemami59)
+
+# Add Support for TypeScript Config Files
+
+## Summary
+
+Add support for TypeScript config files (`eslint.config.ts`, `eslint.config.mts`, `eslint.config.cts`)
+
+<!-- One-paragraph explanation of the feature. -->
+
+## Motivation
+
+<!-- Why are we doing this? What use cases does it support? What is the expected
+outcome? -->
+
+The primary motivation for adding support for TypeScript configuration files to ESLint is to enhance the developer experience and accommodate the evolving JavaScript ecosystem. As TypeScript's popularity continues to grow, more projects are adopting TypeScript not only for their source code but also for their configuration files. This shift is driven by TypeScript's ability to provide compile-time type checks and IntelliSense. By supporting `eslint.config.ts`, `eslint.config.mts`, and `eslint.config.cts`, ESLint will offer first-class support to TypeScript users, allowing them to leverage these benefits directly within their ESLint configuration.
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+
+The goal is to seamlessly support TypeScript configuration files in ESLint. To achieve this, ESLint will need to recognize and parse TypeScript configuration files in the same way it does for JavaScript configuration files. This will involve creating the configuration file resolution logic to recognize `.ts`, `.mts`, and `.cts` files as valid ESLint configuration files. We will need to treat `eslint.config.mts` files as ECMAScript Modules (ESM) and `eslint.config.cts` files as CommonJS (CJS). The module resolution for `eslint.config.ts` files will depend on the `type` field in the nearest `package.json`. The maintainers of ESLint have raised some valid concerns some of which include:
```

Comment:
Again, it's possible that I'm overlooking something but it seems that Jiti makes no difference between `.ts`, `.mts` and `.cts` files, as these are all transpiled to CommonJS, for example:

```ts
export default [];

console.log(typeof module);      // prints "object"
console.log(typeof require);     // prints "function"
console.log(typeof exports);     // prints "object"
console.log(typeof __filename);  // prints "string"
console.log(typeof __dirname);   // prints "string"
```

The output is consistent with CommonJS, and it doesn't matter if this file is named `eslint.config.ts`, `eslint.config.mts` or `eslint.config.cts`.

If my analysis is not correct, could you add a reproducible example to show the differences between the file formats?

Reply:
> Again, it's possible that I'm overlooking something but it seems that Jiti makes no difference between `.ts`, `.mts` and `.cts` files, as these are all transpiled to CommonJS, for example:
> 
> ```ts
> export default [];
> 
> console.log(typeof module);      // prints "object"
> console.log(typeof require);     // prints "function"
> console.log(typeof exports);     // prints "object"
> console.log(typeof __filename);  // prints "string"
> console.log(typeof __dirname);   // prints "string"
> ```
> 
> The output is consistent with CommonJS, and it doesn't matter if this file is named `eslint.config.ts`, `eslint.config.mts` or `eslint.config.cts`.
> 
> If my analysis is not correct, could you add a reproducible example to show the differences between the file formats?

You are correct, `jiti` handles everything synchronously hence why it doesn't support top level `await`. It transforms everything to CommonJS which means users can use either `module.exports` or `export default` and that level of interoprability, is exactly why `jiti` is so widely used _specifically_ for handling config files.